skbio/stats/composition.py

```diff
@@ -2058,3 +2061,423 @@ def dirmult_ttest(
+
+    # Test if metadata has missing values
+    if metadata.isnull().values.any():
+        raise ValueError("Cannot handle missing values in metadata.")
+
+    # Test if metadata and data have the same index, regardless of order
+    if not data.index.sort_values().equals(metadata.index.sort_values()):
+        raise ValueError("Data and metadata must have the same index.")
+
+    rng = get_rng(seed)
+
+    posterior = [
+        rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+    ]
+
+    dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
+
+    _call_res = _lme_call(
+        formula=formula,
+        table=dir_table,
+        metadata=metadata,
+        groups=groups,
+        reml=reml,
+        method=method,
+        fit_kwargs=fit_kwargs,
+        **kwargs,
+    )
+
+    res = _call_res[0]
+    list_of_vars = _call_res[2]
+
+    # Creating an empty dict to store sum of values (Using a DataFrame threw errors)
+    # uses a separate array for each covariate and each feature
+    # array index: 0: pvalue, 1: CI(2.5), 2: CI(97.5), 3: log2fc, 4: qvalue
+    res_dict = {}
+
+    for feature in data.columns:
+        res_ = res_dict[feature] = {}
+        for covar in list_of_vars:
+            res_[covar] = [0] * 5
+
+    for single_covar_data in res:
+        group = res_dict[single_covar_data["FeatureID"]][single_covar_data["Covariate"]]
+        for i, key in enumerate(("pvalue", "CI(2.5)", "CI(97.5)", "Log2(FC)")):
+            group[i] = single_covar_data[key]
+
+    for i in range(1, draws):
+        posterior = [
+            rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+        ]
+        dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
+
+        _call_ires = _lme_call(
+            formula=formula,
+            table=dir_table,
+            metadata=metadata,
+            groups=groups,
+            reml=reml,
+            method=method,
+            fit_kwargs=fit_kwargs,
+            **kwargs,
+        )
+
+        ires = _call_ires[0]
+
+        if ires is None:
+            continue
+
+        # online average to avoid holding all of the results in memory
+        for single_covar_data in ires:
+            curr_feature_id = single_covar_data["FeatureID"]
+            curr_covariate = single_covar_data["Covariate"]
+            res_ = res_dict[curr_feature_id][curr_covariate]
+            res_[0] = (i * res_[0] + single_covar_data["pvalue"]) / (i + 1)
+            res_[1] = np.minimum(res_[1], single_covar_data["CI(2.5)"])
+            res_[2] = np.maximum(res_[2], single_covar_data["CI(97.5)"])
+            res_[3] = (i * res_[3] + single_covar_data["Log2(FC)"]) / (i + 1)
+
+    # convert all log fold changes to base 2
+    log2_ = np.log(2)
+    for feature in data.columns:
+        for covar in list_of_vars:
+            res_ = res_dict[feature][covar]
+            for i in range(1, 4):
```

Comment:
I think these relate to the indices above. But if the data are a common type, and the operation to perform is common, why not store in a contiguous `np.array` or `DataFrame` and apply in bulk?